design/SOURCE&DESTINATION.md

```diff
@@ -0,0 +1,203 @@
+
+```kotlin
+abstract class Source {
+    abstract val cancelCause: Throwable?
+
+    abstract fun receive(): Buffer
+
+    abstract suspend fun awaitBytes(): Boolean
+
+    abstract fun cancel(cause: Throwable = CancellationException("Source has been cancelled"))
+}
+```
+
+## Destination
+
+To handle different outputs there is another abstraction: `Destination`.
+You can write `Buffer` using the `write` method:
+
+```kotlin
+val destination = mySocket.destination
+
+val count = destination.write(buffer)
+println("$count bytes written")
+```
+
+The `write` call writes as much data (between `buffer.readPosition` and `buffer.writePosition`) as it can to the
+destination. It modifies `buffer.readPosition` and returns amount of written bytes.
+
+While the `Source` is responsible for creating `Buffer`, `Destination` is responsible for releasing it. If the `buffer`
+has no bytes available after write, it will be automatically released.
+
+How could you write a whole `Buffer` to the `Destination`? The spin on the `write` method is not a good idea, so there
+is a method `awaitFreeSpace()` that suspends until the destination has some free space to write:
+
+```kotlin
+suspend fun Destination.writeBuffer(buffer: Buffer) {
+    while (true) {
+        write(buffer)
+
+        if (buffer.isEmpty()) return
+        awaitFreeSpace()
+    }
+}
+```
+
+If `Destination` has some buffered bytes, they should be flushed in `awaitFreeSpace`.
+
+### How to end the conversation?
+
+After writing The `close()` method terminates `Destination`. If there are some pending bytes they will be sent:
+
+```kotlin
+suspend fun Destination.writeTrailer(trailer: ByteArray) {
+    writeFully(trailer)
+    close()
+}
+```
+
+You can pass close cause as exception to the close method. It will be stored to the `closedCause` and available for
+tracing. There is also `isClosed` flag to check if the destination is already closed.
+
+### Flush
+
+It's hard to write human-readable justification of having flush method without introducing a new abstractions, so I will
+keep it as TODO("Rewrite this").
+
+If you have chained transformations on the `Destination` like:
+
+```kotlin
+val destination = mySocket.destination
+    .compressed()
+    .buffered()
+```
+
+and a single `message`. There is a common pattern when you want to send this message and make sure that all data is
+actually written to file or socket(for instance it can be a http message header). Some transformations could keep an
+internal buffer because of efficiency reasons.
+
+To make sure that all data is written to the destination, you can call `flush()`. The `flush()` call suspends until all
+data is actually written to the final destination:
+
+```kotlin
+destination.writeFully(message)
+destination.flush()
+// At this point message should be compressed and written to the socket.
+```
+
+## Final Listing
+
+```kotlin
+abstract class Destination {
+    abstract val isClosed: Boolean
```

Comment:
What's the practical use case for `isClosed`? Can we get rid of it same as we removed `isCancelled`/`canRead` in `Source`?